### PR TITLE
Adding ocf-url tests

### PR DESCRIPTION
**Pull request by iherman**
Adding simple tests that check the handling of relative URL-s. All three refer to an image to be displayed via an `img` tag.

Note that two out of the three tests _are not valid_ in terms of epubcheck, but they should still be properly handled by the Reading System per the defined algorithm.

---

 @dlazin, do we want to think about another metadata tag marking the test as 'invalid EPUB, but that is all right'?

### CODE REVIEW

**Review by rdeltour — APPROVED** [review]
Looks good to me! (caveat: I didn't check the test metadata as I'm not familiar with the test suite requirements)

**Comment by dlazin** [comment]
> @dlazin, do we want to think about another metadata tag marking the test as 'invalid EPUB, but that is all right'?

Yes, I think we should either do that or manually note it in the test description.

File: tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en">
```
Comment (dlazin) [inline tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml:1]: Remember, only ever one underscore test names. The underscore is the separator between the identifier for the normative statement (hyphenated) and the optional identifier for separate tests for the same normative statement (also hyphenated, if multi-word).

So perhaps ocf-url_relative-leak? (I think "link" is implicit, but up to you.)

File: tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        text-align: center
+      }
+      img {
+        width: 60%;
+      }
+      figcaption {
+        font-size: smaller;
+        font-style: italic;
+      }
+    </style>
```
Comment (dlazin) [inline tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml:15]: This CSS is fine but seems unnecessary; won't affect the test, so I would omit.

File: tests/ocf-url_link_leaking_relative/EPUB/package.opf [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:epub="http://www.idpf.org/2007/ops" version="3.0" xml:lang="en" unique-identifier="pub-id">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Open Container Format</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-03-06</dc:date>
+    <dc:description>Use a relative link with several double-dot path segments from the content to a photograph. The folder hierarchy containing the photograph starts at the root level; the number of segments exceeds the height of the folder hierarchy.</dc:description>
+    <dc:identifier id="pub-id">ocf-url_link_leaking_relative</dc:identifier>
+    <dc:language>en</dc:language>
+    <dc:title>Potentially leaking relative link</dc:title>
```
Comment (dlazin) [inline tests/ocf-url_link_leaking_relative/EPUB/package.opf:10]: Perhaps "Relative image reference exceeds depth of hierarchy"? I think it's clearer, from that, what this test is (if you haven't read the markup).

File: tests/ocf-url_link_path_absolute/EPUB/package.opf [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:epub="http://www.idpf.org/2007/ops" version="3.0" xml:lang="en" unique-identifier="pub-id">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Open Container Format</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-03-01</dc:date>
+    <dc:description>Use a path-absolute link from the content to a photograph. The folder hierarchy containing the photograph starts at the root level.</dc:description>
```
Comment (dlazin) [inline tests/ocf-url_link_path_absolute/EPUB/package.opf:7]: For clarity, I would say "beginning with a leading slash" or similar.

File: tests/ocf-url_link_path_absolute/EPUB/package.opf [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    <dc:identifier id="pub-id">ocf-url_link_path_absolute</dc:identifier>
+    <dc:language>en</dc:language>
+    <dc:title>Path-absolute links</dc:title>
```
Comment (dlazin) [inline tests/ocf-url_link_path_absolute/EPUB/package.opf:10]: Perhaps "Absolute image reference with leading slash"?

File: tests/ocf-url_link_path_absolute/EPUB/package.opf [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:epub="http://www.idpf.org/2007/ops" version="3.0" xml:lang="en" unique-identifier="pub-id">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Open Container Format</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-03-01</dc:date>
```
Comment (dlazin) [inline tests/ocf-url_link_path_absolute/EPUB/package.opf:6]: Nit: weird that this has a different date from the other test in this PR (but acceptable).

File: tests/ocf-url_link_path_absolute/EPUB/content_001.xhtml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en">
```
Comment (dlazin) [inline tests/ocf-url_link_path_absolute/EPUB/content_001.xhtml:1]: Perhaps ocf-url_absolute?

File: tests/ocf-url_link_relative/EPUB/package.opf [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:epub="http://www.idpf.org/2007/ops" version="3.0" xml:lang="en" unique-identifier="pub-id">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Open Container Format</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-03-05</dc:date>
```
Comment (dlazin) [inline tests/ocf-url_link_relative/EPUB/package.opf:6]: Same, date differs from other tests. (Optional.)

File: tests/ocf-url_link_relative/EPUB/package.opf [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    <dc:description>A simple relative link from the content to a photograph. The folder hierarchy containing the photograph starts at the root level.</dc:description>
+    <dc:identifier id="pub-id">ocf-url_link_relative</dc:identifier>
+    <dc:language>en</dc:language>
+    <dc:title>Simple relative link</dc:title>
```
Comment (dlazin) [inline tests/ocf-url_link_relative/EPUB/package.opf:10]: Relative image reference within depth of hierarchy?

File: tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml (hunk 1 above)
Comment (iherman) [reply tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml:1]: This time I would prefer to keep as is. We have a series of tests with 'link', but we have their counterparts with 'parse' in #133.

Unfortunately, the statements that are tested here are relatively convoluted, hence these multiple underscore characters.

I take your point on the usage of underscore vs. dash, so I will rename things to ocf-url_link-relative-leak (and will do the same with the other PR-s.)

File: tests/ocf-url_link_relative/EPUB/content_001.xhtml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en">
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/ocf-url_link_relative/EPUB/content_001.xhtml:1]: See my comment on the renaming

File: tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml (hunk 2 above)
Comment (iherman) [reply tests/ocf-url_link_leaking_relative/EPUB/content_001.xhtml:15]: ok

File: tests/ocf-url_link_leaking_relative/EPUB/package.opf [hunk 11]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:epub="http://www.idpf.org/2007/ops" version="3.0" xml:lang="en" unique-identifier="pub-id">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Open Container Format</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-03-06</dc:date>
+    <dc:description>Use a relative link with several double-dot path segments from the content to a photograph. The folder hierarchy containing the photograph starts at the root level; the number of segments exceeds the height of the folder hierarchy.</dc:description>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/ocf-url_link_leaking_relative/EPUB/package.opf:7]: pk

File: tests/ocf-url_link_leaking_relative/EPUB/package.opf (hunk 3 above)
Comment (iherman) [reply tests/ocf-url_link_leaking_relative/EPUB/package.opf:10]: ok. I was struggling with this one.

File: tests/ocf-url_link_path_absolute/EPUB/package.opf (hunk 6 above)
Comment (iherman) [reply tests/ocf-url_link_path_absolute/EPUB/package.opf:6]: Yeah... those dates are not used anywhhere:-) But I changed it.

File: tests/ocf-url_link_path_absolute/EPUB/package.opf (hunk 4 above)
Comment (iherman) [reply tests/ocf-url_link_path_absolute/EPUB/package.opf:7]: I added the term, but kept the official things as well, which refers back to the specification text.

File: tests/ocf-url_link_path_absolute/EPUB/package.opf (hunk 5 above)
Comment (iherman) [reply tests/ocf-url_link_path_absolute/EPUB/package.opf:10]: I prefer to keep the official term, which is used in the spec.

File: tests/ocf-url_link_relative/EPUB/content_001.xhtml (hunk 10 above)
[the remark this answers is not in the document]
Comment (iherman) [reply tests/ocf-url_link_relative/EPUB/content_001.xhtml:1]: See above.

**Pull request merged**